framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,439 @@
+                        printWriter.print(basetypes.get(aMethod.returnType.description.toString()));
+                    } else {
+                        printWriter.print("java.lang.Object");
+                    }
+                    printWriter.print(" ");
+                }
+                printWriter.print(methodName);
+                printWriter.print("(");
+                boolean firstParam = true;
+                if (!aMethod.receiver.type.tlAnnotationsHere.isEmpty()
+                        || !aMethod.receiver.type.innerTypes.isEmpty()) {
+                    // Only output the receiver if there is something to
+                    // say.  This is a bit inconsistent with the return
+                    // type, but so be it.
+                    printAField(aMethod.receiver, "this", classname);
+                    firstParam = false;
+                }
+                for (Integer index : aMethod.parameters.keySet()) {
+                    if (!firstParam) {
+                        printWriter.print(", ");
+                    }
+                    AField param = aMethod.parameters.get(index);
+                    // AMethod doesn't actually track real parameter names.
+                    // Fortunately, the stub file parser also doesn't, so
+                    // this code can safely ignore that problem and use a generic
+                    // name instead.
+                    //
+                    // TODO: use the actual parse tree of the method
+                    // to figure out the real parameter names, and then thread them
+                    // through to here.
+                    printAField(param, "param" + index, classname);
+                    firstParam = false;
+                }
+                printWriter.println(");");
+            }
+            for (int i = 0; i < curlyCount; i++) {
+                printWriter.println("}");
+            }
+        }
+    }
+
+    /**
+     * Prints the type parameters of the class given by the fully-qualified name passes as {@code
+     * classname}.
+     */
+    private void printTypeParameters(String classname) {
+        TypeElement type = types.get(classname);
+        if (type == null) {
+            return;
+        }
+        List<? extends TypeParameterElement> typeParameters = type.getTypeParameters();
+        if (typeParameters == null || typeParameters.isEmpty()) {
+            return;
+        }
+        printWriter.print("<");
+        boolean first = true;
+        for (TypeParameterElement t : typeParameters) {
+            if (!first) {
+                printWriter.print(", ");
+            }
+            printWriter.print(t.getSimpleName().toString());
+            first = false;
+        }
+        printWriter.print(">");
+    }
+
+    /**
+     * Private constructor that initializes the printWriter and copies over relevant inputs to the
+     * fields in which they are stored.
+     */
+    private SceneToStubWriter(
```

Comment:
Please move the constructor and the static methods below to the top of the file. (After the fields.)